### ISSUE TITLE
[BUG] Zero Offload Is Significantly Slower Than Normal Training

### ISSUE THREAD

**Issue by stanleyshly** [microsoft/DeepSpeed#2998]
I'm training Flan-t5-large(760M parameters) with DeepSpeed . I'm running RoCM(RX 5600XT, 8GB of Vram) with 16GB of Ram backing it on Ubuntu 22.04.

I'm using huggingface for simplicity. 
First, I attempted to train with fp16 on bare hardware without DeepSpeed (the model barely fit and OOMed after a dozen steps).
` examples/pytorch/translation/run_translation.py --model_name_or_path google/flan-t5-large --per_device_train_batch_size 1 --output_dir output_dir --overwrite_output_dir --fp16 --do_train --max_train_samples 500 --num_train_epochs 1 --dataset_name wmt16 --dataset_config "ro-en" --source_lang en --target_lang ro
` I got roughly 3.9 it/seconds.

Then I enabled DeepSpeed zero 1 with this config; I'm leaving only stuff relating to offloading:
```
{
    "fp16": {
        "enabled": "auto",
        "loss_scale": 0,
        "loss_scale_window": 1000,
        "initial_scale_power": 16,
        "hysteresis": 2,
        "min_loss_scale": 1
    },

    "zero_optimization": {
        "stage": 1,
        "offload_optimizer": {
            "device": "cpu",
            "pin_memory": true
        },
        "allgather_partitions": true,
        "allgather_bucket_size": 2e8,
        "overlap_comm": true,
        "reduce_scatter": true,
        "reduce_bucket_size": 2e8,
        "contiguous_gradients": true
    },
}
```
This led to a hug performance hit down 1.3 seconds per iteration. I know that I can't expect linear scaling with offloading, but getting 20% of the performance feels like way too large of a performance loss. What can I adjust to increase the performance with Zero offloading?

**Comment by tjruwase** [comment]
@stanleyshly, how many GPUs are running with? 

Yes, offloading could be costly, but 5X slowdown does seem too high. Two things to try: 
1. Enable `wall_clock_breakdown` to further understand what is going on. 
2. Increase the batch size to improve the throughput.

**Comment by stanleyshly** [comment]
Only one gpu, which could explain why(isn't DeepSpeed designed for dozens to hundreds of GPUs?)
I tried your second suggestion first, and wow, increasing batch size really work well.
| Batch Size| seconds/iteration |
| ------------- | ------------- |
|  No Deepspeed |  0.256  |
|  1 |  1.3 |
|  2 |  1.1-1.44   |
|  4  |   1.2-1.7  |
|  8  |  1.17-1.4   |
|  16 |   1.387-1.76 |
 At a batch size, the equivalent throughput is faster than without deepspeed. 

In addition, here is my wall_clock_breakdown for a batch size of 1:

```
[2023-03-13 13:32:09,981] [INFO] [logging.py:77:log_dist] [Rank 0] rank=0 time (ms) | forward_microstep: 104.78 | backward_microstep: 934.26 | backward_inner_microstep: 83.01 | backward_allreduce_microstep: 851.22 | step_microstep: 1.27
[2023-03-13 13:32:09,981] [INFO] [logging.py:77:log_dist] [Rank 0] rank=0 time (ms) | forward: 104.78 | backward: 934.26 | backward_inner: 83.01 | backward_allreduce: 851.22 | step: 1.27
```
And here is the trace for batch size of 16:
```
[2023-03-13 13:30:26,678] [INFO] [logging.py:77:log_dist] [Rank 0] rank=0 time (ms) | forward_microstep: 262.76 | backward_microstep: 1055.67 | backward_inner_microstep: 188.92 | backward_allreduce_microstep: 866.73 | step_microstep: 1.28
[2023-03-13 13:30:26,678] [INFO] [logging.py:77:log_dist] [Rank 0] rank=0 time (ms) | forward: 262.76 | backward: 1055.67 | backward_inner: 188.92 | backward_allreduce: 866.73 | step: 1.28
```

I can't tell what the bottleneck is, but backward_microsteps looks quite similar?

**Comment by tjruwase** [comment]
> Only one gpu, which could explain why(isn't DeepSpeed designed for dozens to hundreds of GPUs?)

Yes, the ZeRO optimizations are essentially memory deduplication for Data Parallel training and are thus effective on more than one GPU. 

In this case you are using the offloading mechanism, which incurs overheads from GPU<->CPU copies and CPU-based optimizer execution. The benefit is the reduction in GPU memory consumption, and so scaling the batch size is a way to counteract the overheads. 

The wall clock breakdown looks reasonable to me. Are you okay with closing this issue? Thanks!

**Comment by stanleyshly** [comment]
Yes, looks like I solved the issue.
